### ISSUE TITLE
[BUG] incorrect z movements with arcs

### ISSUE THREAD

**Issue by flaviut** [MarlinFirmware/Marlin#24041]
### Did you test the latest `bugfix-2.0.x` code?

Yes, and the problem still exists.

### Bug Description

I'm slicing this part in vase mode in Cura, with [Arc Welder](https://github.com/FormerLurker/ArcWelderLib) enabled and "Allow 3d Arcs" on. Turning 3d arcs off works fine.

Here's a video. At first, it prints back & forth around the circle without a problem (or maybe just a little bit of a problem). Then it switches into going all the way around continuously, but the z-axis does not move. Eventually it decides to move the Z-axis 2-3mm, all at once.

I looked through the g-code, and I don't see any rapid z-axis moves like this.

https://user-images.githubusercontent.com/5213469/163416610-50612416-52d0-4408-b6b7-c6ed366db945.mp4

### Bug Timeline

Unknown, I've on used bugfix-2.0.x

### Expected behavior

I expected it to print smoothly in vase mode.

### Version of Marlin Firmware

29d3f6a024672a960f7e033e360178f4e80ce98b

### Printer model

Ender 5

### Electronics

SKR2

### Bed Leveling

UBL

### Your Slicer

Cura

### Host Software

Octoprint

### Additional information & file uploads

config: https://github.com/MarlinFirmware/Marlin/compare/bugfix-2.0.x...flaviut:ender5-skr2-bugfix-2.0.x

log output from around the right place:
<details>
```
Recv: ok N1042 P0 B2
Send: N1056 G3 X120.578 Y98.430 Z4.107 I-3.951 J15.283 E299.15126*37
Recv: echo:N1043 G2 X119.255 Y123.242 Z3.906 I27.314 J25.938 E295.30732*14
Recv: ok N1043 P0 B2
Send: N1057 G3 X124.946 Y102.407 Z4.123 I-198.887 J222.828 E299.46074*37
Recv: echo:N1044 G1 X117.888 Y124.749 Z3.912 E295.41391*12
Recv: ok N1044 P0 B2
Send: N1058 G3 X126.477 Y107.529 Z4.138 I-4.076 J4.008 E299.75215*46
Recv: echo:N1045 G3 X115.357 Y126.305 Z3.920 I-4.144 J-3.905 E295.57138*4
Recv: ok N1045 P0 B2
Send: N1059 G3 X125.175 Y113.912 Z4.156 I-266.760 J-51.089 E300.09344*63
Recv: echo:N1046 G3 X112.482 Y126.407 Z3.928 I-1.647 J-5.845 E295.72357*12
Recv: ok N1046 P0 B2
Send: N1060 G3 X121.568 Y120.577 Z4.177 I-15.192 J-3.913 E300.49437*50
Recv: echo:N1047 G3 X106.087 Y125.166 Z3.946 I34.633 J-195.565 E296.06487*21
Recv: ok N1047 P0 B2
Send: N1061 G3 X117.362 Y125.142 Z4.194 I-156.046 J-139.554 E300.81959*8
Recv: echo:N1048 G3 X99.419 Y121.573 Z3.967 I4.032 J-15.467 E296.46544*45
Recv: ok N1048 P0 B2
Send: N1062 G3 X115.362 Y126.303 Z4.200 I-3.842 J-4.314 E300.94163*6
Recv: echo:N1049 G3 X95.023 Y117.626 Z3.983 I192.530 J-218.853 E296.77496*16
Recv: ok N1049 P0 B2
Send: N1063 G3 X112.149 Y126.437 Z4.209 I-1.844 J-5.625 E301.11225*13
Recv: echo:N1050 G3 X93.570 Y112.478 Z3.998 I4.146 J-3.949 E297.06637*28
Recv: ok N1050 P0 B2
Send: N1064 G3 X105.583 Y125.037 Z4.227 I23.128 J-124.567 E301.46401*30
Recv: echo:N1051 G3 X94.830 Y106.089 Z4.016 I234.577 J42.943 E297.40754*7
Recv: ok N1051 P0 B2
Send: N1065 G3 X99.219 Y121.381 Z4.247 I4.464 J-15.138 E301.85201*47
Recv: echo:N1052 G3 X98.428 Y99.421 Z4.037 I15.378 J3.993 E297.80829*54
Recv: ok N1052 P0 B2
Send: N1066 G3 X94.769 Y117.220 Z4.264 I66.676 J-75.767 E302.17124*30
Recv: echo:N1053 G3 X102.391 Y95.038 Z4.053 I256.793 J228.203 E298.11785*49
Recv: ok N1053 P0 B2
Send: N1067 G3 X93.540 Y112.144 Z4.279 I4.348 J-3.740 E302.45522*20
Recv: echo:N1054 G3 X107.525 Y93.546 Z4.068 I3.979 J4.113 E298.40917*63
Recv: ok N1054 P0 B2
Send: N1068 G3 X94.966 Y105.583 Z4.297 I114.230 J21.392 E302.80702*12
Recv: echo:N1055 G3 X113.911 Y94.828 Z4.086 I-51.296 J272.067 E298.75040*46
Recv: ok N1055 P0 B2
Send: N1069 G3 X98.620 Y99.220 Z4.317 I15.124 J4.454 E303.19495*63
Recv: echo:N1056 G3 X120.578 Y98.430 Z4.107 I-3.951 J15.283 E299.15126*37
Recv: ok N1056 P0 B2
Send: N1070 G3 X102.794 Y94.785 Z4.334 I98.130 J88.173 E303.51407*62
Recv: echo:N1057 G3 X124.946 Y102.407 Z4.123 I-198.887 J222.828 E299.46074*37
Recv: ok N1057 P0 B2
Send: N1071 G3 X107.861 Y93.517 Z4.349 I3.777 J4.335 E303.79817*63
Recv: echo:N1058 G3 X126.477 Y107.529 Z4.138 I-4.076 J4.008 E299.75215*46
Recv: ok N1058 P0 B2
Send: N1072 M117 ETL 00h04m01s*62
Recv: echo:N1059 G3 X125.175 Y113.912 Z4.156 I-266.760 J-51.089 E300.09344*63
Recv:  T:245.14 /245.00 B:71.00 /70.00 @:91 B@:22
Recv: ok N1059 P0 B2
Send: N1073 M73 P14*87
Recv: echo:N1060 G3 X121.568 Y120.577 Z4.177 I-15.192 J-3.913 E300.49437*50
Recv: ok N1060 P0 B2
Send: N1074 G3 X114.416 Y94.968 Z4.367 I-23.089 J119.839 E304.14995*47
Recv: echo:N1061 G3 X117.362 Y125.142 Z4.194 I-156.046 J-139.554 E300.81959*8
Recv: ok N1061 P0 B2
Send: N1075 G3 X120.779 Y98.621 Z4.387 I-4.469 J15.153 E304.53790*44
Recv: echo:N1062 G3 X115.362 Y126.303 Z4.200 I-3.842 J-4.314 E300.94163*6
Recv: ok N1062 P0 B2
Send: N1076 G3 X122.846 Y100.557 Z4.395 I-32.660 J36.942 E304.68628*34
Recv: echo:N1063 G3 X112.149 Y126.437 Z4.209 I-1.844 J-5.625 E301.11225*13
Recv: ok N1063 P0 B2
Send: N1077 G3 X125.200 Y102.808 Z4.404 I-26.510 J30.080 E304.85696*32
Recv: echo:N1064 G3 X105.583 Y125.037 Z4.227 I23.128 J-124.567 E301.46401*30
Recv: ok N1064 P0 B2
Send: N1078 G3 X126.506 Y107.865 Z4.419 I-4.247 J3.794 E305.14114*44
Recv: echo:N1065 G3 X99.219 Y121.381 Z4.247 I4.464 J-15.138 E301.85201*47
Recv: ok N1065 P0 B2
Send: N1079 G3 X125.904 Y110.631 Z4.426 I-40.673 J-7.404 E305.28945*54
Recv: echo:N1066 G3 X94.769 Y117.220 Z4.264 I66.676 J-75.767 E302.17124*30
Recv: ok N1066 P1 B2
Send: N1080 G3 X124.482 Y115.986 Z4.442 I-28.370 J-4.667 E305.58019*61
Recv: echo:N1067 G3 X93.540 Y112.144 Z4.279 I4.348 J-3.740 E302.45522*20
Recv: ok N1067 P0 B2
Send: N1081 G3 X121.378 Y120.779 Z4.457 I-14.873 J-6.230 E305.88100*48
Recv: echo:N1068 G3 X94.966 Y105.583 Z4.297 I114.230 J21.392 E302.80702*12
Recv: ok N1068 P0 B2
Send: N1082 G3 X119.433 Y122.836 Z4.465 I-28.537 J-25.035 E306.02932*11
Recv: echo:N1069 G3 X98.620 Y99.220 Z4.317 I15.124 J4.454 E303.19495*63
Recv: ok N1069 P0 B2
Send: N1083 G3 X116.894 Y125.425 Z4.475 I-23.403 J-20.412 E306.21944*5
Recv: echo:N1070 G3 X102.794 Y94.785 Z4.334 I98.130 J88.173 E303.51407*62
Recv: ok N1070 P0 B2
Send: N1084 G3 X115.256 Y126.330 Z4.480 I-3.274 J-3.992 E306.31796*15
Recv: echo:N1071 G3 X107.861 Y93.517 Z4.349 I3.777 J4.335 E303.79817*63
Recv: ok N1071 P0 B2
Send: N1085 G3 X111.852 Y126.475 Z4.489 I-1.939 J-5.490 E306.49901*13
Recv: echo:N1072 M117 ETL 00h04m01s*62
Recv: //action:notification ETL 00h04m01s
Recv: ok N1072 P0 B3
Send: N1086 G1 X106.169 Y125.187 Z4.505 E306.80426*1
Recv: echo:N1073 M73 P14*87
Recv: ok N1073 P0 B3
Recv: echo:N1074 G3 X114.416 Y94.968 Z4.367 I-23.089 J119.839 E304.14995*47
Send: N1087 G3 X99.017 Y121.191 Z4.528 I3.888 J-15.357 E307.23831*35
Recv: ok N1074 P0 B2
Send: N1088 G1 X94.925 Y117.266 Z4.543 E307.53534*55
Recv: echo:N1075 G3 X120.779 Y98.621 Z4.387 I-4.469 J15.153 E304.53790*44
Recv: ok N1075 P0 B2
Send: N1089 G3 X93.501 Y111.852 Z4.559 I4.146 J-3.985 E307.84144*30
Recv: echo:N1076 G3 X122.846 Y100.557 Z4.395 I-32.660 J36.942 E304.68628*34
Recv: ok N1076 P0 B2
Send: N1090 G1 X94.809 Y106.17 Z4.575 E308.14688*8
Recv: echo:N1077 G3 X125.200 Y102.808 Z4.404 I-26.510 J30.080 E304.85696*32
Recv: ok N1077 P0 B2
Send: N1091 G3 X98.811 Y99.018 Z4.598 I15.371 J3.905 E308.58109*62
Recv: echo:N1078 G3 X126.506 Y107.865 Z4.419 I-4.247 J3.794 E305.14114*44
Recv:  T:245.43 /245.00 B:70.94 /70.00 @:83 B@:0
Recv: ok N1078 P0 B2
Send: N1092 G1 X102.75 Y94.942 Z4.613 E308.87802*1
Recv: echo:N1079 G3 X125.904 Y110.631 Z4.426 I-40.673 J-7.404 E305.28945*54
Recv: ok N1079 P0 B2
Send: N1093 G3 X108.158 Y93.479 Z4.629 I4.013 J4.106 E309.18451*54
Recv: echo:N1080 G3 X124.482 Y115.986 Z4.442 I-28.370 J-4.667 E305.58019*61
Recv: ok N1080 P0 B2
Send: N1094 G1 X113.83 Y94.807 Z4.645 E309.48967*13
Recv: echo:N1081 G3 X121.378 Y120.779 Z4.457 I-14.873 J-6.230 E305.88100*48
Recv: ok N1081 P0 B2
Send: N1095 G3 X120.981 Y98.812 Z4.668 I-3.894 J15.339 E309.92395*39
Recv: echo:N1082 G3 X119.433 Y122.836 Z4.465 I-28.537 J-25.035 E306.02932*11
Recv: ok N1082 P0 B2
Send: N1096 G1 X125.042 Y102.765 Z4.683 E310.22083*15
Recv: echo:N1083 G3 X116.894 Y125.425 Z4.475 I-23.403 J-20.412 E306.21944*5
Recv: ok N1083 P0 B2
Send: N1097 G3 X126.545 Y108.157 Z4.699 I-4.069 J4.040 E310.52715*33
Recv: echo:N1084 G3 X115.256 Y126.330 Z4.480 I-3.274 J-3.992 E306.31796*15
Recv: ok N1084 P0 B2
Send: N1098 G1 X125.195 Y113.831 Z4.715 E310.83268*6
Recv: echo:N1085 G3 X111.852 Y126.475 Z4.489 I-1.939 J-5.490 E306.49901*13
Recv: ok N1085 P0 B2
Send: N1099 G3 X121.187 Y120.981 Z4.738 I-15.355 J-3.909 E311.26707*50
Recv: echo:N1086 G1 X106.169 Y125.187 Z4.505 E306.80426*1
Recv: ok N1086 P0 B2
Send: N1100 G3 X117.096 Y125.147 Z4.754 I-119.442 J-113.200 E311.57293*10
Recv: echo:N1087 G3 X99.017 Y121.191 Z4.528 I3.888 J-15.357 E307.23831*35
Recv: ok N1087 P0 B2
Send: N1101 G3 X115.207 Y126.343 Z4.760 I-3.941 J-4.134 E311.69082*9
Recv: echo:N1088 G1 X94.925 Y117.266 Z4.543 E307.53534*55
Recv: ok N1088 P0 B2
Send: N1102 G3 X111.568 Y126.504 Z4.770 I-2.067 J-5.518 E311.88476*3
Recv: echo:N1089 G3 X93.501 Y111.852 Z4.559 I4.146 J-3.985 E307.84144*30
Recv: ok N1089 P0 B2
Send: N1103 G3 X105.136 Y124.893 Z4.788 I28.582 J-127.760 E312.23216*27
Recv: echo:N1090 G1 X94.809 Y106.17 Z4.575 E308.14688*8
Recv: ok N1090 P0 B3
Send: N1104 G3 X98.824 Y120.993 Z4.809 I4.949 J-15.067 E312.62447*34
Recv: echo:N1091 G3 X98.811 Y99.018 Z4.598 I15.371 J3.905 E308.58109*62
Recv: ok N1091 P0 B2
Send: N1105 G3 X94.543 Y116.703 Z4.825 I71.797 J-75.927 E312.94202*21
Recv: echo:N1092 G1 X102.75 Y94.942 Z4.613 E308.87802*1
Recv: ok N1092 P0 B3
Send: N1106 G3 X93.473 Y111.562 Z4.840 I4.490 J-3.616 E313.22758*30
Recv: echo:N1093 G3 X108.158 Y93.479 Z4.629 I4.013 J4.106 E309.18451*54
Recv: ok N1093 P0 B2
Send: N1107 G3 X95.108 Y105.136 Z4.858 I114.190 J25.633 E313.57500*15
Recv: echo:N1094 G1 X113.83 Y94.807 Z4.645 E309.48967*13
Recv: ok N1094 P0 B3
Send: N1108 G3 X99.008 Y98.825 Z4.879 I15.004 J4.911 E313.96731*61
Recv: echo:N1095 G3 X120.981 Y98.812 Z4.668 I-3.894 J15.339 E309.92395*39
Recv: ok N1095 P0 B2
Send: N1109 G3 X103.311 Y94.560 Z4.895 I115.699 J112.426 E314.28475*58
Recv: echo:N1096 G1 X125.042 Y102.765 Z4.683 E310.22083*15
Recv: ok N1096 P0 B2
Send: N1110 G3 X108.444 Y93.451 Z4.910 I3.646 J4.443 E314.57039*51
Recv: echo:N1097 G3 X126.545 Y108.157 Z4.699 I-4.069 J4.040 E310.52715*33
Recv:  T:245.31 /245.00 B:70.79 /70.00 @:87 B@:0
Recv: ok N1097 P0 B2
Send: N1111 G3 X114.864 Y95.108 Z4.928 I-34.163 J145.628 E314.91779*46
Recv: echo:N1098 G1 X125.195 Y113.831 Z4.715 E310.83268*6
Recv: ok N1098 P0 B3
Send: N1112 G3 X121.174 Y99.009 Z4.948 I-4.916 J15.006 E315.31011*44
Recv: echo:N1099 G3 X121.187 Y120.981 Z4.738 I-15.355 J-3.909 E311.26707*50
Recv: ok N1099 P0 B2
Send: N1113 G3 X125.423 Y103.325 Z4.965 I-90.293 J93.141 E315.62745*41
Recv: echo:N1100 G3 X117.096 Y125.147 Z4.754 I-119.442 J-113.200 E311.57293*10
Recv: ok N1100 P0 B2
Send: N1114 G3 X126.572 Y108.450 Z4.980 I-4.402 J3.678 E315.91318*42
Recv: echo:N1101 G3 X115.207 Y126.343 Z4.760 I-3.941 J-4.134 E311.69082*9
Recv: ok N1101 P0 B2
Send: N1115 G3 X125.872 Y111.191 Z4.988 I-43.152 J-9.560 E316.06139*63
Recv: echo:N1102 G3 X111.568 Y126.504 Z4.770 I-2.067 J-5.518 E311.88476*3
Recv: ok N1102 P0 B2
Send: N1116 G3 X124.267 Y116.483 Z5.003 I-28.696 J-5.814 E316.35156*48
Recv: echo:N1103 G3 X105.136 Y124.893 Z4.788 I28.582 J-127.760 E312.23216*27
Recv: ok N1103 P0 B2
Send: N1117 G3 X120.990 Y121.174 Z5.018 I-14.553 J-6.676 E316.65298*61
Recv: echo:N1104 G3 X98.824 Y120.993 Z4.809 I4.949 J-15.067 E312.62447*34
Recv: ok N1104 P0 B2
Send: N1118 G3 X118.974 Y123.160 Z5.026 I-27.151 J-25.545 E316.80125*3
Recv: echo:N1105 G3 X94.543 Y116.703 Z4.825 I71.797 J-75.927 E312.94202*21
Recv: ok N1105 P1 B2
Send: N1119 G3 X116.370 Y125.638 Z5.036 I-22.238 J-20.762 E316.98970*5
Recv: echo:N1106 G3 X93.473 Y111.562 Z4.840 I4.490 J-3.616 E313.22758*30
Recv: ok N1106 P0 B2
Send: N1120 G3 X115.063 Y126.368 Z5.040 I-3.369 J-4.497 E317.06835*0
Recv: echo:N1107 G3 X95.108 Y105.136 Z4.858 I114.190 J25.633 E313.57500*15
Recv: ok N1107 P0 B2
Send: N1121 G3 X110.991 Y126.462 Z5.051 I-2.168 J-5.658 E317.28577*8
Recv: echo:N1108 G3 X99.008 Y98.825 Z4.879 I15.004 J4.911 E313.96731*61
Recv: ok N1108 P0 B2
Send: N1122 G3 X105.025 Y124.854 Z5.068 I44.248 J-176.040 E317.60949*17
Recv: echo:N1109 G3 X103.311 Y94.560 Z4.895 I115.699 J112.426 E314.28475*58
Recv: ok N1109 P1 B2
Send: N1123 G3 X98.420 Y120.590 Z5.090 I5.133 J-15.199 E318.02552*36
Recv: echo:N1110 G3 X108.444 Y93.451 Z4.910 I3.646 J4.443 E314.57039*51
Recv: ok N1110 P0 B2
Send: N1124 G3 X94.561 Y116.588 Z5.105 I156.158 J-154.440 E318.31678*30
Recv: echo:N1111 G3 X114.864 Y95.108 Z4.928 I-34.163 J145.628 E314.91779*46
Recv: ok N1111 P0 B2
Send: N1125 G3 X93.514 Y110.985 Z5.121 I4.501 J-3.740 E318.62843*21
Recv: echo:N1112 G3 X121.174 Y99.009 Z4.948 I-4.916 J15.006 E315.31011*44
Recv: ok N1112 P0 B2
Send: N1126 G3 X95.142 Y105.029 Z5.138 I209.576 J54.085 E318.95191*4
Recv: echo:N1113 G3 X125.423 Y103.325 Z4.965 I-90.293 J93.141 E315.62745*41
Recv: ok N1113 P2 B2
Send: N1127 G3 X99.404 Y98.428 Z5.160 I15.200 J5.138 E319.36769*48
Recv: echo:N1114 G3 X126.572 Y108.450 Z4.980 I-4.402 J3.678 E315.91318*42
Recv:  T:245.21 /245.00 B:70.59 /70.00 @:89 B@:1
Recv: ok N1114 P0 B2
Send: N1128 G3 X103.425 Y94.576 Z5.175 I191.654 J196.038 E319.65940*61
Recv: echo:N1115 G3 X125.872 Y111.191 Z4.988 I-43.152 J-9.560 E316.06139*63
Recv: ok N1115 P0 B2
Send: N1129 G3 X109.021 Y93.491 Z5.191 I3.765 J4.447 E319.97123*60
Recv: echo:N1116 G3 X124.267 Y116.483 Z5.003 I-28.696 J-5.814 E316.35156*48
Recv: ok N1116 P0 B2
Send: N1130 G3 X114.976 Y95.143 Z5.208 I-59.637 J226.532 E320.29500*35
Recv: echo:N1117 G3 X120.990 Y121.174 Z5.018 I-14.553 J-6.676 E316.65298*61
Recv: ok N1117 P0 B2
Send: N1131 G3 X121.571 Y99.405 Z5.230 I-5.108 J15.137 E320.71052*33
Recv: echo:N1118 G3 X118.974 Y123.160 Z5.026 I-27.151 J-25.545 E316.80125*3
Recv: ok N1118 P0 B2
Send: N1132 G3 X125.637 Y103.715 Z5.246 I-131.615 J128.237 E321.02098*36
Recv: echo:N1119 G3 X116.370 Y125.638 Z5.036 I-22.238 J-20.762 E316.98970*5
Recv: ok N1119 P0 B2
Send: N1133 G3 X126.532 Y109.027 Z5.261 I-4.625 J3.511 E321.31423*34
Recv: echo:N1120 G3 X115.063 Y126.368 Z5.040 I-3.369 J-4.497 E317.06835*0
Recv: ok N1120 P0 B2
Send: N1134 G3 X124.861 Y114.972 Z5.278 I-265.307 J-71.364 E321.63775*58
Recv: echo:N1121 G3 X110.991 Y126.462 Z5.051 I-2.168 J-5.658 E317.28577*8
Recv: ok N1121 P0 B2
Send: N1135 G3 X120.594 Y121.571 Z5.300 I-15.116 J-5.095 E322.05362*55
Recv: echo:N1122 G3 X105.025 Y124.854 Z5.068 I44.248 J-176.040 E317.60949*17
Recv: ok N1122 P0 B2
Send: N1136 G1 X118.736 Y123.321 Z5.307 E322.18733*5
Recv: echo:N1123 G3 X98.420 Y120.590 Z5.090 I5.133 J-15.199 E318.02552*36
Recv: ok N1123 P0 B2
Send: N1137 G1 X118.54 Y123.523 E322.20207*111
Recv: echo:N1124 G3 X94.561 Y116.588 Z5.105 I156.158 J-154.440 E318.31678*30
Recv: ok N1124 P0 B2
Send: N1138 G3 X116.275 Y125.623 Z5.316 I-23.910 J-23.517 E322.36391*12
Recv: echo:N1125 G3 X93.514 Y110.985 Z5.121 I4.501 J-3.740 E318.62843*21
Recv: ok N1125 P0 B2
Send: N1139 G3 X114.985 Y126.379 Z5.320 I-3.429 J-4.373 E322.44246*13
Recv: echo:N1126 G3 X95.142 Y105.029 Z5.138 I209.576 J54.085 E318.95191*4
Recv: ok N1126 P0 B2
Send: N1140 G3 X111.125 Y126.593 Z5.331 I-2.233 J-5.356 E322.64885*11
Recv: echo:N1127 G3 X99.404 Y98.428 Z5.160 I15.200 J5.138 E319.36769*48
Recv: ok N1127 P0 B2
Send: N1141 G3 X104.686 Y124.736 Z5.349 I28.885 J-112.249 E322.99996*27
Recv: echo:N1128 G3 X103.425 Y94.576 Z5.175 I191.654 J196.038 E319.65940*61
Recv: ok N1128 P0 B2
Send: N1142 G3 X98.435 Y120.596 Z5.370 I5.469 J-15.047 E323.39634*39
Recv: echo:N1129 G3 X109.021 Y93.491 Z5.191 I3.765 J4.447 E319.97123*60
Recv: ok N1129 P0 B2
Send: N1143 G3 X94.332 Y116.175 Z5.386 I159.132 J-151.800 E323.71236*28
Recv: echo:N1130 G3 X114.976 Y95.143 Z5.208 I-59.637 J226.532 E320.29500*35
Recv: ok N1130 P0 B2
Send: N1144 G3 X93.385 Y111.119 Z5.401 I4.578 J-3.474 E323.99167*30
Recv: echo:N1131 G3 X121.571 Y99.405 Z5.230 I-5.108 J15.137 E320.71052*33
Recv: ok N1131 P0 B2
Send: N1145 G3 X95.263 Y104.686 Z5.419 I197.128 J54.057 E324.34278*1
Recv: echo:N1132 G3 X125.637 Y103.715 Z5.246 I-131.615 J128.237 E321.02098*36
Recv:  T:245.08 /245.00 B:70.39 /70.00 @:92 B@:3
Recv: ok N1132 P0 B2
Send: N1146 G3 X99.407 Y98.438 Z5.440 I14.951 J5.418 E324.73923*52
Recv: echo:N1133 G3 X126.532 Y109.027 Z5.261 I-4.625 J3.511 E321.31423*34
Recv: ok N1133 P0 B2
Send: N1147 G3 X103.839 Y94.350 Z5.456 I147.089 J155.020 E325.05513*59
Recv: echo:N1134 G3 X124.861 Y114.972 Z5.278 I-265.307 J-71.364 E321.63775*58
Recv: ok N1134 P0 B2
Send: N1148 G3 X108.886 Y93.362 Z5.471 I3.511 J4.551 E325.33438*54
Recv: echo:N1135 G3 X120.594 Y121.571 Z5.300 I-15.116 J-5.095 E322.05362*55
Recv: ok N1135 P0 B2
Send: N1149 G3 X115.315 Y95.263 Z5.489 I-46.389 J168.704 E325.68563*40
Recv: echo:N1136 G1 X118.736 Y123.321 Z5.307 E322.18733*5
Recv: ok N1136 P0 B2
Send: N1150 G3 X121.560 Y99.410 Z5.510 I-5.394 J14.899 E326.08212*34
Recv: echo:N1137 G1 X118.54 Y123.523 E322.20207*111
Recv: ok N1137 P0 B3
Send: N1151 G3 X125.632 Y103.852 Z5.526 I-114.448 J109.002 E326.39784*38
Recv: echo:N1138 G3 X116.275 Y125.623 Z5.316 I-23.910 J-23.517 E322.36391*12
Recv: ok N1138 P1 B2
Send: N1152 G3 X126.660 Y108.892 Z5.541 I-4.523 J3.547 E326.67715*39
Recv: echo:N1139 G3 X114.985 Y126.379 Z5.320 I-3.429 J-4.373 E322.44246*13
Recv: ok N1139 P0 B2
Send: N1153 G3 X124.545 Y115.835 Z5.560 I-120.044 J-32.775 E327.05748*48
Recv: echo:N1140 G3 X111.125 Y126.593 Z5.331 I-2.233 J-5.356 E322.64885*11
```
</details>

GCode: [gcode.zip](https://github.com/MarlinFirmware/Marlin/files/8486254/gcode.zip)

**Comment by ellensp** [comment]
"Allow 3D Arcs
This option allows G2/G3 commands to be generated when using vase mode. This is an **experimental** option, and it's possible that there are some unknown firmware issues when adding Z coordinates to arc commands. That being said, I've gotten pretty good results from this option. At some point, this will be enabled by default.

Type: Flag
Default: Disabled
Short Parameter: -z
Long Parameter: --allow-3d-arcs
Example: ArcWelder "C:\thing.gcode" --allow-3d-arcs"

**Comment by ManuelMcLure** [comment]
An arc with a Z component should work per the code in `plan_arc`:
```Plan an arc in 2 dimensions, with linear motion in the other axes.```
So if this is not working it’s a bug.

**Comment by ManuelMcLure** [comment]
That said, you should examine the gcode to make sure the G2/G3 commands are as expected. This might be an ArcWelder bug.

**Comment by ellensp** [comment]
It not a bug when it is not implemented feature. 
Official gcode specification  https://reprap.org/wiki/G-code#G2_.26_G3:_Controlled_Arc_Move
No Z component

**Comment by ManuelMcLure** [comment]
Yes, but Marlin _does_ implement it.

**Comment by DerAndere1** [comment]
Marlin documentation already mentions the Z component: https://marlinfw.org/docs/gcode/G002-G003.html

The Z axis, is mapped to `axis_l` in G2_G3.cpp. Last time I tested it, it worked as expected with additional axes. The attached G code contains G3 commands with Z component as well as G3 commands without Z component. So it is not clear if this is a bug or not. We need to test isolated G3 commands with Z component.

**Comment by DerAndere1** [comment]
@flaviut please test the gcode script attached to this comment. It is an excerpt from the example gcode you provided with all extrusion removed and feedrate reduced. It should home, then move the toolhead in a helix, starting at hight Z1.964 and ending at Z2.240 . On my machine (`BOARD_ANET_10`, A4988 stepper drivers) it works without issues. In Configuration_adv.h I have
```
#define ARC_SUPPORT                   // Requires ~3226 bytes
#if ENABLED(ARC_SUPPORT)
  #define MIN_ARC_SEGMENT_MM      0.1 // (mm) Minimum length of each arc segment
  #define MAX_ARC_SEGMENT_MM      1.0 // (mm) Maximum length of each arc segment
  #define MIN_CIRCLE_SEGMENTS    72   // Minimum number of segments in a complete circle
  //#define ARC_SEGMENTS_PER_SEC 50   // Use the feedrate to choose the segment length
  #define N_ARC_CORRECTION       25   // Number of interpolated segments between corrections
  //#define ARC_P_CIRCLES             // Enable the 'P' parameter to specify complete circles
  //#define SF_ARC_FIX                // Enable only if using SkeinForge with "Arc Point" fillet procedure
#endif
```

[CB_issue_24041_test1.zip](https://github.com/MarlinFirmware/Marlin/files/8535715/CB_issue_24041_test1.zip)

**Comment by DerAndere1** [comment]
Does the issue also occur if you disable bed leveling?

**Comment by flaviut** [comment]
My printer is in bits right now :)

I'll take a look and let you know as soon as it's back together

**Comment by jars28** [comment]
I did a bit of testing. Will test the script when I'm back at my printer. I made a log of my attempts and added comments.
[marlin 2.1.x console session w notes.txt](https://github.com/MarlinFirmware/Marlin/files/8927544/marlin.2.1.x.console.session.w.notes.txt)
.

**Comment by DerAndere1** [comment]
In the test script are you sure that the following sequence should result in any movement? I don't think so:
```
Send: G91
Recv: ok
[...]
Send: G18
Recv: ok
[...]

//HAPPY LIL VERT CIRCLE
Send: G2 X5 Z5 I5
```
Try this instead (I get an arc in the x-y plane with simultaneous z movement (i.e. helical movement, as expected):
```
G91
G18
G2 X5 Z5 I2.5
```

Otherwise I did not observe any unexpected behaviour when running your test script. Some of your annotations indicate that  you have wrong expectations what the G2/G3 commands do. Or your machine indeed behaves totally different from mine.

**Comment by jars28** [comment]
It moved in a vertical circle in the X and Z directions, with G91 active and after G18 was entered, but when I added a Y coordinate, there was no movement in the Y direction. My expectation from the documentation on G2 and G3 was that it would calculate and segment the linear Y positions along with calculating the positions along the arc for the two relevant axis(es?). 

The entire function looks like it has been rewritten in the bugfix version since I tested it, so I will try again when I am at the office with the printer and down time and not out in the field.

**Comment by DerAndere1** [comment]
I confused G18 and M18 and forgot to enable CNC_WORKSPACE_PLANES. Padon me. I will redo the test, too

**Comment by DerAndere1** [comment]
With `CNC_WORKSPACE_PLANES` enabled, I did not observe any unexpected behaviour with the test from https://github.com/MarlinFirmware/Marlin/issues/24041#issuecomment-1158838657. I.e., where the comment in the script says "HAPPY LIL VERT CIRCLE, Y DOESN'T MOVE AT ALL", I actually observed a helical movement (vertical circle in XZ plane with simultaneous movement of the Y axis). Where the comments in the script say "TINY ARC, NO Z MOVE", I observed movement of Z for the small distance of 0.1 mm. Again, I was testing with all bed leveling disabled in my configs
